tests/test_data_loader.py

```diff
@@ -498,6 +499,35 @@ def test_set_epoch_in_batch_sampler(self):
         dataloader.set_epoch(1)
         assert batch_sampler.epoch == 1
 
+    def test_ensure_dataloader_gets_cleaned_up(self):
+        # Ensure that the dataloader gets cleaned up properly
+        class Dummy:
+            def __init__(self):
+                dataset = list(range(16))
+                dataloader = DataLoader(dataset, batch_size=4)
+
+                self.accelerator = Accelerator()
+                self.dataloader = self.accelerator.prepare_data_loader(dataloader)
+
+                self.iter = iter(self.dataloader)
+
+            def __call__(self, *args, **kwds):
+                return next(self.iter)
+
+        instance = Dummy()
+        assert instance().tolist() == [0, 1, 2, 3]
+
+        # Create weak references to the objects that *should* be cleaned up if the instance is deleted
```

Comment:
I wonder if we can have a more realistic unit test that is closer to the failure cases you presented and that doesn't  rely on `weakref` to check if the object was cleaned up.

Reply:
As a heads up, the `gc.get_objects()` call -> filtering is comparatively very slow. Still definitely <1s, but it's not really recommended. That's why I went with `weakref`.

I'm totally fine with switching back to `gc.get_objects()` to make sure `weakref` isn't doing something weird, but the test case might be slowed down a bit (again: only a tiny bit I believe).

Reply:
I've had people yell at us about `gc` stuff in transformers quite a bit, perfectly okay doing weakref here instead :sweat:

Reply:
Let's keep it as-is then. I think this is the cleanest way to track whether an object is GC'd.